File: factory/src/adapters/artifact/artifact-hash.ts
/**
 * Content addressing for artifacts.
 *
 * Uses only the standard Node `node:crypto` module so the operational bundle
 * keeps its `node:*`-only dependency invariant.
 */

import { createHash, randomUUID } from 'node:crypto'

/** Content address prefix used by every artifact hash. */
export const ARTIFACT_HASH_PREFIX = 'sha256'

/** Computes the content address (`sha256:<hex>`) of a payload. */
export function computeArtifactHash(data: Buffer | Uint8Array): string {
  return `${ARTIFACT_HASH_PREFIX}:${createHash('sha256').update(data).digest('hex')}`
}

/** Generates a fresh, opaque artifact identifier. */
export function createArtifactId(): string {
  return randomUUID()
}

File: factory-service/src/main/kotlin/io/whozoss/factory/artifact/domain/ArtifactHash.kt
package io.whozoss.factory.artifact.domain

import java.security.MessageDigest
import java.util.UUID

/**
 * Content addressing for artifacts. Ported from
 * `factory/src/adapters/artifact/artifact-hash.ts`.
 */
object ArtifactHash {

    /** Content address prefix used by every artifact hash. */
    const val PREFIX = "sha256"

    /** Computes the content address (`sha256:<hex>`) of a payload. */
    fun compute(data: ByteArray): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(data)
        return "$PREFIX:${digest.toHex()}"
    }

    /** Generates a fresh, opaque artifact identifier. */
    fun createArtifactId(): String = UUID.randomUUID().toString()

    /** Extracts the bare hex digest from a `sha256:<hex>` content address. */
    fun digest(hash: String): String =
        if (hash.startsWith("$PREFIX:")) hash.substring(PREFIX.length + 1) else hash

    private fun ByteArray.toHex(): String {
        val out = StringBuilder(size * 2)
        for (byte in this) {
            out.append(HEX[(byte.toInt() shr 4) and 0x0F])
            out.append(HEX[byte.toInt() and 0x0F])
        }
        return out.toString()
    }

    private val HEX = "0123456789abcdef".toCharArray()
}
